download.py

```diff
@@ -17,55 +20,102 @@
 @dataclass
 class DownloadResult:
     file_path: str  # target_file_path,  # TODO harmonize with dane-download-worker
-    download_time: float = -1  # time (secs) taken to receive data after request
-    mime_type: str = "unknown"  # download_data.get("mime_type", "unknown"),
+    mime_type: str
+    provenance: dict
+    download_time: float = -1  # time (ms) taken to receive data after request
+    error: str = ""
     content_length: int = -1  # download_data.get("content_length", -1),
 
 
-def download_uri(uri: str) -> Optional[DownloadResult]:
+def download_uri(uri: str) -> DownloadResult:
     logger.info(f"Trying to download {uri}")
     if validate_s3_uri(uri):
         logger.info("URI seems to be an s3 uri")
         return s3_download(uri)
-    return http_download(uri)
+    if validate_http_uri(uri):
+        logger.info("URI seems to be an HTTP URI")
+        return http_download(uri)
+    return DownloadResult(
+        uri, "", dict(), -1, "Input failure: URI is neither S3, nor HTTP"
+    )
 
 
 def http_download(url: str) -> DownloadResult:
     logger.info(f"Checking if {url} was already downloaded")
+    start_time = time.time()
+
+    provenance = {
+        "activity_name": "Input download",
+        "activity_description": "Downloads the input file from INPUT_URI",
+        "processing_time_ms": -1,
+        "start_time_unix": start_time,
+        "parameters": [],
+        "software_version": "",
+        "input_data": url,
+        "output_data": "",
+        "steps": [],
+    }
+
     fn = os.path.basename(urlparse(url).path)
-    input_file = os.path.join(input_file_dir, fn)
-    # TODO rethink how to return mime_type
-    _, extension = get_asset_info(input_file)
+    asset_id, extension = get_asset_info(fn)
+
+    input_file_dir = os.path.join(data_base_dir, asset_id)
```

Comment:
You define this here, but you want it to be the same as the output_dir defined elsewhere. So it would be better to create directory and pass it to the download_uri function as an argument

Reply:
I prefer having the `download_uri` function do all the download and input-related logic and keep the `run` function in `run_pipeline` as simple as possible (only focus on getting the right input+output from each step and passing it along)